Went to the simplest approach.

ManOnTheMountainTech opened the pull request:
- Look for intellimouse taillight arrival
- Start a worker thread to thunk to pageable code
- Set the taillight black
- Rebased with main

forderud commented on TailLight/device.h:
```diff
@@ -1,13 +1,37 @@
 
+#include "debug.h"
+
+#define MSINTELLIMOUSE_USBINTERFACE5_PREFIX L"\\??\\HID#VID_045E&PID_082A&MI_01&Col05"
```
I would prefer to avoid hardcoding the HW ID here, since it will need to be manually kept in sync with the INX file. Would it be possible for you to instead retrieve it using `IoGetDevicePropertyData` or similar as described in https://learn.microsoft.com/en-us/windows-hardware/drivers/install/hardware-ids ?

forderud commented on TailLight/device.h:
```diff
@@ -1,13 +1,37 @@
+
+#define BEGIN_WITH(x) { \
+        auto &_ = x;
+#define END_WITH() }
```
What are these defines used for?

ManOnTheMountainTech replied:
This was future work for dealing with members of a structure in a row, in order to avoid "I Love to Type" (ILT). Other languages have a With() statement, like:

BEGIN_WITH(GPIO_PORT_CONTROL)
    _.AFSEL = 0
    _.DIGITAL_ENABLE = 1
    _.PIN_ENABLE = 0x5
END_WITH

However, the final version that I'm going with won't need to work with members of a structure or class in succession, so I removed it. Again, learning the checkin rules of the organization.

forderud replied:
Please remove unrelated changes from this pull request. I only want to merge the changes needed to fix the reported problem.

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,250 @@
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+        NukeWdfHandle<WDFIOTARGET>(target);
```
I don't think you should delete `target` here, since the object is passed as an input argument to the function. Please instead move the `NukeWdfHandle` call to the client code, and call it if `TryToOpenIoTarget` fails.

ManOnTheMountainTech replied:
Done. Now in SetBlackAsync(). I'm taking more breaks to prevent such errors.

forderud replied:
What do you mean by "done"? The pull request have not yet been updated to address the comment.

forderud commented on TailLight/driver.h:
```diff
@@ -12,9 +12,15 @@
 /** Memory allocation tag name (for debugging leaks). */
-static constexpr ULONG POOL_TAG = 'ffly';
+static constexpr ULONG POOL_TAG = 'ylff';
+static constexpr ULONG WDF_POOL_TAG = 'wTLT'; // Taillight WPF allocated
```
Comments:
* Please don't bundle changes to `POOL_TAG` with the other changes in this PR. Submit it as a separate PR if you want to change it.
* Why are you introducing `WDF_POOL_TAG`? It doesn't seem to be used anywhere?

ManOnTheMountainTech replied:
Done. I now know the rules. WDF_POOL_TAG is to be used in a future checkin so that WDF allocations on behalf of TailLight will be traceable to taillight.. Otherwide, a generic pool tag is used that could refer to any WDF driver.

forderud replied:
Sorry, but the new `POOL_TAG` define is still in your PR, so the review comment is not yet "done".

forderud commented on TailLight/debug.h:
```diff
@@ -0,0 +1,10 @@
+
+#pragma once
+
+#define TRACE_FN_ENTRY KdPrintEx((DPFLTR_IHVDRIVER_ID, DPFLTR_TRACE_LEVEL, "TailLight: Entry %s\n", __func__));
```
Please don't bundle changes to driver tracing/logging with the other changes in this PR. Feel free to instead submit it in a _separate_ PR if you want changes in this area.

ManOnTheMountainTech replied:
I put a #pragma message and a TODO in to remove this before checkin.

forderud replied:
I don't see any changes. The `TRACE_FN_ENTRY` define is stil present in this pull request.

forderud replied:
Thanks.

ManOnTheMountainTech commented:
This is my interpretation of what the documentation says to do at https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfiotarget/nf-wdfiotarget-wdfiotargetopen.

 

“If a call to WdfIoTargetOpen fails, the driver should call WdfObjectDelete <https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfobject/nf-wdfobject-wdfobjectdelete>  to delete the I/O target object.”

 

From: Fredrik Orderud ***@***.***> 
Sent: Thursday, February 22, 2024 2:38 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Assign ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

@forderud commented on this pull request.

  _____  

In TailLight/SetTaillightBlack.cpp <https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1499031509> :

> +        &openParams,
+        &DeviceContext.PdoName,
+        FILE_WRITE_ACCESS);
+
+    openParams.ShareAccess = FILE_SHARE_WRITE | FILE_SHARE_READ;
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    // Ensure freed if fails.
+    status = WdfIoTargetOpen(target, &openParams);
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx((DPFLTR_IHVDRIVER_ID,
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+        NukeWdfHandle<WDFIOTARGET>(target);

I don't think you should delete target here, since the object is passed as an input argument to the function. Please instead move the NukeWdfHandle call to the client code, and call it if TryToOpenIoTarget fails.

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#pullrequestreview-1895506017> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV72R7FAJDQ6BSL7BBGLYU4NXRAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQOJVGUYDMMBRG4> .
You are receiving this because you were assigned.  <https://github.com/notifications/beacon/AGBDV75I4JTLAZR7JWFKRYTYU4NXRA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTTQ7MQGC.gif> Message ID: ***@***.*** ***@***.***> >

forderud commented:
> This is my interpretation of what the documentation says to do at https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfiotarget/nf-wdfiotarget-wdfiotargetopen. “If a call to WdfIoTargetOpen fails, the driver should call WdfObjectDelete <https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfobject/nf-wdfobject-wdfobjectdelete> to delete the I/O target object.”

I fully agree with you on that. You should still delete the `target` object if `WdfIoTargetOpen` fail. My request was only that you should move the code that performs the deletion to the calling function.

ManOnTheMountainTech commented:
WDF assigns a generic pool tag to allocations unless overridden. I was probably getting around to setting the Intellimouse WDF pool tag but never completed the task. The goal was to be able to narrow down leaky code by specific pool tags. In any case it’s removed.

 

From: Fredrik Orderud ***@***.***> 
Sent: Thursday, February 22, 2024 2:40 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Assign ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

@forderud commented on this pull request.

  _____  

In TailLight/driver.h <https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1499035171> :

>  /** Memory allocation tag name (for debugging leaks). */
-static constexpr ULONG POOL_TAG = 'ffly';
+static constexpr ULONG POOL_TAG = 'ylff';
+static constexpr ULONG WDF_POOL_TAG = 'wTLT'; // Taillight WPF allocated

Comments:

*	Please don't bundle changes to POOL_TAG with the other changes in this PR. Submit it as a separate PR if you want to change it.
*	Why are you introducing WDF_POOL_TAG? It doesn't seem to be used anywhere?

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#pullrequestreview-1895511696> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV7532JM5VGCD5AS2HATYU4OBZAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQOJVGUYTCNRZGY> .
You are receiving this because you were assigned.  <https://github.com/notifications/beacon/AGBDV73PW4QWO2L3B6W53CDYU4OBZA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTTQ7M3JA.gif> Message ID: ***@***.*** ***@***.***> >

ManOnTheMountainTech commented:
I had set it for all Taillight WDF allocations. In any case, for another PR.

 

From: Fredrik Orderud ***@***.***> 
Sent: Thursday, February 22, 2024 2:40 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Assign ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

@forderud commented on this pull request.

  _____  

In TailLight/driver.h <https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1499035171> :

>  /** Memory allocation tag name (for debugging leaks). */
-static constexpr ULONG POOL_TAG = 'ffly';
+static constexpr ULONG POOL_TAG = 'ylff';
+static constexpr ULONG WDF_POOL_TAG = 'wTLT'; // Taillight WPF allocated

Comments:

*	Please don't bundle changes to POOL_TAG with the other changes in this PR. Submit it as a separate PR if you want to change it.
*	Why are you introducing WDF_POOL_TAG? It doesn't seem to be used anywhere?

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#pullrequestreview-1895511696> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV7532JM5VGCD5AS2HATYU4OBZAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQOJVGUYTCNRZGY> .
You are receiving this because you were assigned.  <https://github.com/notifications/beacon/AGBDV73PW4QWO2L3B6W53CDYU4OBZA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTTQ7M3JA.gif> Message ID: ***@***.*** ***@***.***> >

forderud commented:
> WDF assigns a generic pool tag to allocations unless overridden. I was probably getting around to setting the Intellimouse WDF pool tag but never completed the task. The goal was to be able to narrow down leaky code by specific pool tags. In any case it’s removed.

I still see the pool tag change in this PR. You'll need to (force) push your changes if you want to update the PR on the server.

ManOnTheMountainTech commented:
Symbolic links are supposed to be opaque. Parsing the string representation of them into a hardware ID is not good.

 

First, get our hardware ID and USB configuration.

 

USB IO targets can have their interfaces and endpoints selected. I’m working on getting the USB device descriptor for the device that is represented by the symbolic link. Everything’s parsed out this way and defined by an open spec.

Next, configure the IO target’s interface and endpoint to match our endpoint.

Use that io target for the SetBlack() call.

 

I could also get the hardware ID from the device represented by the symbolic link, and match to see if it’s our hardware ID.

 

From: Fredrik Orderud ***@***.***> 
Sent: Monday, February 26, 2024 10:11 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

@ManOnTheMountainTech <https://github.com/ManOnTheMountainTech>  Any progress on updating this PR?

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#issuecomment-1964813243> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV77IHEMVO3C6WWMVKO3YVTF2PAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRUHAYTGMRUGM> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AGBDV7ZX4QFF372EARSEBJDYVTF2PA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVDSV3W.gif> Message ID: ***@***.*** ***@***.***> >

forderud commented:
Gentle reminder.

ManOnTheMountainTech commented:
I’ve been working on it yesterday and today.

 

The PnP callback is only being called once on another old system that I set up. The symbolic link returned is failing to open. On another machine, OpenTarget is working as there are more callbacks.

 

From: Fredrik Orderud ***@***.***> 
Sent: Sunday, March 03, 2024 9:44 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

Gentle reminder.

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#issuecomment-1975241848> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV72JGTVVJIOPOYBM3LTYWNOOPAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZVGI2DCOBUHA> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AGBDV7ZBXAETFLNHUHMYDYLYWNOOPA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVXPGHQ.gif> Message ID: ***@***.*** ***@***.***> >

forderud commented:
> I’ve been working on it yesterday and today. The PnP callback is only being called once on another old system that I set up. The symbolic link returned is failing to open. On another machine, OpenTarget is working as there are more callbacks.

That's probably good, but I'm struggling to understand why that prevents you from updating the PR to address the other review comments you received more than a week ago.

ManOnTheMountainTech commented:
Last week had six appointments between both houses and 12 hours of travel
time. In addition, I needed to attend a VP promotion party for my wife at
Microsoft
https://www.linkedin.com/feed/update/urn:li:activity:7169478179806547968/.
Setting up another test machine took time. The login screen came up black
after install which took some time to diagnose. Not attending the VP
promotion event would have been a marriage-testing event. In short, I did
not get all my hours in.

 

Instead of addressing all the issues in the PR at once I'll separate out
what is ready and check those in. I'll be catching up this week and through
the weekend.

 

  _____  

From: Fredrik Orderud ***@***.***>
Sent: Monday, March 4, 2024 1:28:16 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention
***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach.
(PR #54) 

 

I've been working on it yesterday and today. The PnP callback is only being
called once on another old system that I set up. The symbolic link returned
is failing to open. On another machine, OpenTarget is working as there are
more callbacks.

That's probably good, but I'm struggling to understand why that prevents you
from updating the PR to address the other review comments you received more
than a week ago.

-
Reply to this email directly, view it on GitHub
<https://github.com/forderud/IntelliMouseDriver/pull/54#issuecomment-1976110
632> , or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AGBDV74B6HWWD3KRSIAOY73YW
Q5DBAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZWGEYTA
NRTGI> .
You are receiving this because you were mentioned.
<https://github.com/notifications/beacon/AGBDV77F4QAWHOLZULY3LA3YWQ5DBA5CNFS
M6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTT
VZEHCQ.gif> Message ID:
***@***.***>

forderud commented:
> Do you want (a) checkin to occur first, then the questions to the PR answered, or (b) can I answer first, then checkin?

Do as you wish.

ManOnTheMountainTech commented:
I have not pushed anything yet. Testing changes. It is done on my side. I'll switch to commenting after pushing.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Fredrik Orderud ***@***.***>
Sent: Tuesday, March 5, 2024 12:51:40 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)


@forderud commented on this pull request.

________________________________

In TailLight/SetTaillightBlack.cpp<https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1512381609>:

> +        &openParams,
+        &DeviceContext.PdoName,
+        FILE_WRITE_ACCESS);
+
+    openParams.ShareAccess = FILE_SHARE_WRITE | FILE_SHARE_READ;
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    // Ensure freed if fails.
+    status = WdfIoTargetOpen(target, &openParams);
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx((DPFLTR_IHVDRIVER_ID,
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+        NukeWdfHandle<WDFIOTARGET>(target);


What do you mean by "done"? The pull request have not yet been updated to address the comment.

—
Reply to this email directly, view it on GitHub<https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1512381609>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGBDV76KJYUC7T2KT3ZX4UDYWWBRZAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMJWGI3TCMBUGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

forderud commented on TailLight/TailLight.vcxproj:
```diff
@@ -33,6 +33,7 @@
     <DriverType>KMDF</DriverType>
     <PlatformToolset>WindowsKernelModeDriver10.0</PlatformToolset>
     <ConfigurationType>Driver</ConfigurationType>
+    <Driver_SpectreMitigation>Spectre</Driver_SpectreMitigation>
```
Why are you bundling Spectre-mitigation changes with the changes in this PR??  Please submit such change requests in a separate RP.

forderud commented on TailLight/device.h:
```diff
@@ -1,12 +1,29 @@
 
 /** Driver-specific struct for storing instance-specific data. */
-struct DEVICE_CONTEXT {
+typedef struct DEVICE_CONTEXT {
```
Why are you adding a "typedef" prefix here?

forderud commented on TailLight/device.cpp:
```diff
@@ -1,50 +1,186 @@
+        return status;
+    }
+
+    hwId.Buffer = (PWCHAR)WdfMemoryGetBuffer(memory, (SIZE_T*) & (hwId.MaximumLength));
```
Please use a `size_t bufferLength` output variable here for configuring the `Length` and `MaximumLength` members like already done for `DevicePropertyPhysicalDeviceObjectName`.

forderud commented on TailLight/device.cpp:
```diff
@@ -78,49 +216,24 @@ Routine Description:
+    }
 
-        KdPrint(("TailLight: PdoName: %wZ\n", deviceContext->PdoName)); // outputs "\Device\00000083
+    status = AllocateHwIdString(device, pDeviceContext->OurHardwareId);
```
I think you'll need to call `WdfDeviceGetIoTarget(target)` here to get the WDFIOTARGET from WDFDEVICE when calling `AllocateHwIdString`.

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,260 @@
+        
+        UNICODE_STRING theirHwId = {};
+
+        status = AllocateHwIdString(hidTarget, theirHwId);
```
`theirHwId` allocations will here pile up causing leakage if `SetBlackAsync` is called multiple times. Could this become an issue?

ManOnTheMountainTech replied:
No. Parented to to the IO target.

forderud commented on TailLight/SetBlack.h:
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+typedef struct _SET_BLACK_WORKITEM_CONTEXT {
```
Would you mind using C++ style struct declaration, so that it becomes:
```
struct SET_BLACK_WORKITEM_CONTEXT {
    UNICODE_STRING symLink;
};
```

forderud replied:
Gentle reminder.

ManOnTheMountainTech replied:
Sure. Please verify that it's in.

forderud replied:
Thanks for explaining.

forderud commented on TailLight/TailLight.vcxproj:
```diff
@@ -82,12 +83,13 @@ copy "$(WDKBinRoot_x64)\certmgr.exe" $(PackageDir)</Command>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
     <Link>
-      <AdditionalDependencies>%(AdditionalDependencies);$(DDK_LIB_PATH)\hidparse.lib</AdditionalDependencies>
+      <AdditionalDependencies>%(AdditionalDependencies);$(DDK_LIB_PATH)\hidparse.lib;$(DDK_LIB_PATH)\ntstrsafe.lib</AdditionalDependencies>
```
Are you sure that you need to link to `ntstrsafe.lib` here` If yes, then why are you only introducing linking for "debug" builds, and not "release"? This seems very odd to me.

ManOnTheMountainTech replied:
Verified that ntstrsafe.lib does not need to be explicitly brought in.

forderud commented on TailLight/device.cpp:
```diff
@@ -1,50 +1,186 @@
+    return status;
+}
+
+NTSTATUS AllocatePDOName(WDFDEVICE device, UNICODE_STRING& pdoName) 
```
Why are you extracting `DevicePropertyPhysicalDeviceObjectName` retrieval into a separate function here? This doesn't seem to be required for this PR.

ManOnTheMountainTech replied:
Previous work environments had different code quality stages. The idea is that you should do the minimum amount of work to validate a design, then clean up the design afterward it is validated or throw away the code that fails. There has not been broad testing of this code so I submitted the minimum. Copy/paste has low risk and minimum effort.

ManOnTheMountainTech commented:
Vfeaturepp, line 112: WDF_IO_TARGET_OPEN_PARAMS_INIT_OPEN_BY_NAME(&openParams, &deviceContext->PdoName, FILE_WRITE_ACCESS);

 

From: Fredrik Orderud ***@***.***> 
Sent: Wednesday, March 06, 2024 1:35 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

@forderud commented on this pull request.

  _____  

In TailLight/device.cpp <https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1514137717> :

> +    status = IoRegisterPlugPlayNotification(
+        EventCategoryDeviceInterfaceChange,
+        PNPNOTIFY_DEVICE_INTERFACE_INCLUDE_EXISTING_INTERFACES,
+        (PVOID)&GUID_DEVINTERFACE_HID,
+        WdfDriverWdmGetDriverObject(driver),
+        PnpNotifyDeviceInterfaceChange,
+        (PVOID)device,
+        &driverContext->pnpDevInterfaceChangedHandle
+    );
+
+    //TRACE_FN_EXIT;
+
+    return status;
+}
+
+NTSTATUS AllocatePDOName(WDFDEVICE device, UNICODE_STRING& pdoName) 

Why are you extracting DevicePropertyPhysicalDeviceObjectName retrieval into a separate function here? This doesn't seem to be required for this PR.

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#pullrequestreview-1919228033> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV75ICRPRIU36SWH6KWLYW3PL3AVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMJZGIZDQMBTGM> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AGBDV726S5Z35PONTPDHUG3YW3PL3A5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTTSMUMIC.gif> Message ID: ***@***.*** ***@***.***> >

ManOnTheMountainTech commented:
The parent of the allocations is the IO target. The IO Target is explicitly deleted thus deleting the hardware ID at the end of the function. Therefore,

 

1.	If the IO target will not be passed down the stack, the hardware ID allocation will next be deallocated when the IO target is deallocated.
2.	If the IO target will be passed down the stack, then the hardware ID will be taking up unneeded space until SetBlacAsync() completes. In human terms this is very short period of time but would allow a little more memory to be available to the system sooner. So far, I’ve only seen one of the created IO targets be passed down. I felt that there would be two hardware ID deallocation paths to maintain with explicitly deleting the allocated hardware ID for a very small increase in system resources. 

 

Please advise if explicitly deleting the allocated hardware ID would be better.

 

From: Fredrik Orderud ***@***.***> 
Sent: Wednesday, March 06, 2024 1:23 AM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

@forderud commented on this pull request.

  _____  

In TailLight/SetTaillightBlack.cpp <https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1514117115> :

> +        if (!NT_SUCCESS(status)) {
+            KdPrintEx((DPFLTR_IHVDRIVER_ID,
+                DPFLTR_ERROR_LEVEL,
+                "TailLight: 0x%x while creating I/O target from worker thread\n",
+                status));
+            return status;
+        }
+
+        status = TryToOpenIoTarget(hidTarget, symLink);
+    }
+
+    if (NT_SUCCESS(status)) {
+        
+        UNICODE_STRING theirHwId = {};
+
+        status = AllocateHwIdString(hidTarget, theirHwId);

theirHwId allocations will here pile up causing leakage if SetBlackAsync is called multiple times. Could this become an issue?

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#pullrequestreview-1919194793> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV76HOITLVYJGOEN5PV3YW3N5TAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMJZGE4TINZZGM> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AGBDV75BXM2LW4C35A4K7OLYW3N5TA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTTSMSLKS.gif> Message ID: ***@***.*** ***@***.***> >

forderud commented:
> Vfeaturepp, line 112: WDF_IO_TARGET_OPEN_PARAMS_INIT_OPEN_BY_NAME(&openParams, &deviceContext->PdoName, FILE_WRITE_ACCESS);

Sorry, but I don't understand why you're referring to a file that is _not_ changed in this PR. Can you please explain?

forderud commented:
> The parent of the allocations is the IO target. The IO Target is explicitly deleted thus deleting the hardware ID at the end of the function. Therefore,
> ...

Thanks for explaining. Then I agree with the current proposal, since the HWID string will anyhow be deleted as part of the `NukeWdfHandle<WDFIOTARGET>(hidTarget);` call a few lines below.

ManOnTheMountainTech commented:
I read your comments again and got a different perspective. What you’re asking is why I took the code that had not been changed in AddDevice and put it into a function? Because it was easier to understand my own changes and much easier to copy/paste a function to make the new functions so that I could validate my hypothesis. I was creating regressions during development. Previous employment recommended no more than 10 lines per function and thus encapsulating functionality with new functions was required. But I’ll be incorporating #61 which nullifies the entire comment.

 

From: Fredrik Orderud ***@***.***> 
Sent: Wednesday, March 06, 2024 1:36 PM
To: forderud/IntelliMouseDriver ***@***.***>
Cc: Bryan Allen Young ***@***.***>; Mention ***@***.***>
Subject: Re: [forderud/IntelliMouseDriver] Went to the simplest approach. (PR #54)

 

Vfeaturepp, line 112: WDF_IO_TARGET_OPEN_PARAMS_INIT_OPEN_BY_NAME(&openParams, &deviceContext->PdoName, FILE_WRITE_ACCESS);

Sorry, but I don't understand why you're referring to a file that is not changed in this PR. Can you please explain?

—
Reply to this email directly, view it on GitHub <https://github.com/forderud/IntelliMouseDriver/pull/54#issuecomment-1981828918> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGBDV76ICQ6EX73ZONLT4L3YW6D2JAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBRHAZDQOJRHA> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AGBDV75ZBSL4FZ4CDPKRCVDYW6D2JA5CNFSM6AAAAABDUIRGV6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTWEBHTM.gif> Message ID: ***@***.*** ***@***.***> >

forderud commented on TailLight/driver.cpp:
```diff
@@ -11,13 +11,16 @@ NTSTATUS DriverEntry(
 
     WDF_DRIVER_CONFIG params = {};
     WDF_DRIVER_CONFIG_INIT(/*out*/&params, EvtDriverDeviceAdd);
-    params.DriverPoolTag = POOL_TAG;
```
Why did you remove this line?

ManOnTheMountainTech replied:
There was a previous comment that this should be only added in a separate PR. I built a task list and was using e-mail to respond to these issues. I ensured that this was removed as I had a task for it. Hopefully continued practice with rebasing and merging will reduce these kinds of issues. Please verify that it's back in.

forderud replied:
Thanks.

forderud commented on TailLight/driver.cpp:
```diff
@@ -11,13 +11,16 @@ NTSTATUS DriverEntry(
 
+    WDF_OBJECT_ATTRIBUTES attributes = {};
+    WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+    WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&attributes, DRIVER_CONTEXT);
```
Can you please explain the intent of setting "DRIVER_CONTEXT" here?
Also, I think the `WDF_OBJECT_ATTRIBUTES_INIT` call becomes redundant if you're afterwards calling `WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE`.

ManOnTheMountainTech replied:
The callback registered by IoRegisterPlugPlayNotification has to be unregistered when the driver unloads. Otherwise, the driver is called with nothing guaranteed to be present in handling any requests. EvtDriverUnload taked a driver object. Therefore the handle that controls the PlugPlayNotification registration is best stored in the driver objects's context space.

forderud replied:
Thanks for explaining.

forderud commented on TailLight/device.cpp:
```diff
@@ -70,6 +95,62 @@ static UNICODE_STRING GetDevicePropertyString(WDFDEVICE device, DEVICE_REGISTRY_
+    string.Length = (USHORT)bufferLength - sizeof(UNICODE_NULL);
+}
+
+UNICODE_STRING GetDevicePropertyString(WDFIOTARGET target, 
```
It seems like you've introduced a second `GetDevicePropertyString` function that's almost identical to the existing `GetDevicePropertyString` function, with the exception of a `WDFIOTARGET` argument. Can you please try to modify/extend the existing function instead of introducing a semi-duplicated function?

ManOnTheMountainTech replied:
I've fixed using a pointer-to-function approach. This avoids vtables.

forderud replied:
https://github.com/forderud/IntelliMouseDriver/pull/54/files#r1518576669

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,264 @@
+
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+
+        // It's possible to get called twice. Been there, done that?
```
What is meant by `Been there, done that?` here? Please either make the sentence more objective or remove it.

forderud replied:
Gentle reminder.

forderud commented on TailLight/device.cpp:
```diff
@@ -1,60 +1,119 @@
 
     WDFMEMORY memory = 0;
-    NTSTATUS status = WdfDeviceAllocAndQueryProperty(device, DeviceProperty, NonPagedPoolNx, &attributes, &memory);
+    NTSTATUS status = pfnWdfAllocAndQueryTargetPropertyFrom(Source, 
```
Wouldn't it be simpler to instead always call `WdfIoTargetAllocAndQueryTargetProperty` here, instead of using a function pointer?

The `WDFDEVICE`-based client's can then call `WdfDeviceGetIoTarget(device)` to get the local I/O target for the device, so that an `WDFIOTARGET` is always present.

forderud replied:
Alternative proposal that avoids having to introduce a function pointer: https://github.com/forderud/IntelliMouseDriver/pull/63

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,263 @@
+    status = WdfRequestGetStatus(Request);
+    KdPrint(("TailLight: %s WdfRequestSend status: 0x%x\n", __func__, status));
+
+    // One-shot and top of stack, so delete and pray.
```
Why pray here? Isn't this guaranteed to work?

forderud commented on TailLight/device.cpp:
```diff
@@ -1,52 +1,77 @@
+        WdfDriverWdmGetDriverObject(driver),
+        PnpNotifyDeviceInterfaceChange,
+        (PVOID)device,
+        &driverContext->pnpDevInterfaceChangedHandle
```
There seem to be a mismatch with the way `pnpDevInterfaceChangedHandle` is registered & unregistered. It's currently registered when attaching a new _device_ and unregistered when unloading the _driver_. This could cause `pnpDevInterfaceChangedHandle` to be registered twice if inserting two identical IntelliMouse's and only unregistered once after both are disconnected.

Wouldn't it be better to either:
* Move the unregistration code to when the _device_ is removed/disconnected, or
* Move the registration code to when the _driver_ is loaded.

forderud replied:
Thanks for improving on this.

forderud commented on TailLight/device.h:
```diff
@@ -3,11 +3,27 @@
 };
+
+
+template<typename T> inline void NukeWdfHandle(T& handle) {
```
This tiny function is only called from two places. Therefore suggest that you remove it and instead call `WdfObjectDelete(...)` and so on directly from the client code.

forderud commented on TailLight/device.cpp:
```diff
@@ -1,52 +1,77 @@
+
+            // Opening a device may trigger PnP operations. Ensure that either a
+            // timer or a work item is used when opening up a device.
+            // Refer to p356 of Oney and IoGetDeviceObjectPointer.
```
I don't understand what is meant by the "p356 of Oney and IoGetDeviceObjectPointer" comment. Please update the comment to make it clear what is meant.

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,263 @@
+        theirHwId = GetTargetPropertyString(hidTarget,
+            DevicePropertyHardwareID);
+        if (NT_SUCCESS(status)) {
+            if (!RtlEqualUnicodeString(&pDeviceContext->HardwareId,
```
Are you sure that HWID is the "right" identifier to use for matching here? Won't that lead to mix-ups if attaching two identical IntelliMouse's?

I would intuitively think that the PDO name or similar would be a better identifier to use, since identical mouses will still have unique PDO names.

forderud replied:
Thanks for changing the code to use PDO as device identifier.

forderud commented on TailLight/device.cpp:
```diff
@@ -71,6 +95,35 @@ UNICODE_STRING GetTargetPropertyString(WDFIOTARGET target, DEVICE_REGISTRY_PROPE
+    Called near the end of processing an IRP_MN_REMOVE_DEVICE.
+    This work could also be done at EvtDeviceSelfManagedIoCleanup, which
+    comes before this callback. However using the lifetime of the device
+    context means there is little chance of the device context disappearing
```
What is meant by "little chance of the device context disappearing from underneath" here?

Doesn't the framework guarantee that the context will outlive the `EvtCleanupCallback` callback? If not, then where is this WDF limitation documented?

I don't think we should write brittle code that is based on finger crossing and hoping for a good outcome. We should instead utilize the design guarantees offered by WDF & WDM.

ManOnTheMountainTech replied:
This holds as long as WDF & WDM are bug free. I've seen workarounds or fixes implemented for code that's been around for decades.

forderud replied:
Please then either assume that WDF & WDM are bug-free, or add comments links to reported WDF/WDM issues.

forderud replied:
Thanks for removing unreferenced claims from your PR.

ManOnTheMountainTech commented:
The code has been tested on only two machines. Some bug may appear when it
is installed on billions of machines. My confidence is certain that this is
the best approach that I can think of at this time. However, the general
view at Microsoft and Google is that you just don't know what issue will
pop up after code is deployed to the entire world. I can switch to stating
how confident I think the code is at this time. We haven't run a static
analysis which would also help to validate our hypothesis. I suppose I've
turned into a software reliability pessimist.

On Wed, Mar 6, 2024 at 1:20 AM Fredrik Orderud ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In TailLight/device.cpp
> <https://github.com/forderud/IntelliMouseDriver/pull/54#discussion_r1514114335>
> :
>
> >
> -        KdPrint(("TailLight: PdoName: %wZ\n", deviceContext->PdoName)); // outputs "\Device\00000083
> +    status = AllocateHwIdString(device, pDeviceContext->OurHardwareId);
>
> I think you'll need to call WdfDeviceGetIoTarget(target) here to get the
> WDFIOTARGET from WDFDEVICE when calling AllocateHwIdString.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/forderud/IntelliMouseDriver/pull/54#pullrequestreview-1919190493>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGBDV77ECDG67BMFY4WD6ELYW3NWLAVCNFSM6AAAAABDUIRGV6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMJZGE4TANBZGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

forderud commented:
> The code has been tested on only two machines. Some bug may appear when it is installed on billions of machines. My confidence is certain that this is the best approach that I can think of at this time. However, the general view at Microsoft and Google is that you just don't know what issue will pop up after code is deployed to the entire world. I can switch to stating how confident I think the code is at this time. We haven't run a static analysis which would also help to validate our hypothesis. I suppose I've turned into a software reliability pessimist.

Sorry, but this is not a good answer Bryan. I need a better picture of which failures modes could actually occur - and need to be tested & handled, vs. the failure modes that are guaranteed to _never_ occur. How much testing you´ve performed is of little relevance in that regard.

It seems to me, based on reading https://learn.microsoft.com/en-us/windows-hardware/drivers/wdf/framework-object-context-space , that WDF is guaranteeing that the device and context will still be alive when `EvtCleanupCallback` is called. You should therefore remove the "little chance of the device context disappearing from underneath" comment since it´s misleading.

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,268 @@
+
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+
+        // It's possible to get called twice.
```
Please document the situation that could trigger this function to be called twice for a given device?

forderud replied:
Thanks for adding a documentation link to this claim.

forderud commented on TailLight/device.cpp:
```diff
@@ -1,46 +1,70 @@
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION pDevInterface =
+        (PDEVICE_INTERFACE_CHANGE_NOTIFICATION)pvNotificationStructure;
+
+    ASSERT(IsEqualGUID(*(_GUID*)&(pDevInterface->InterfaceClassGuid),
```
I don´t think you need the `*(_GUID*)&` casting here. Same also apply to the IsEqualGUID call below.

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,268 @@
+        }
+    }
+
+ExitAndFree:
```
Please use the `WDFIOTARGET_Wrap` RAII wrapper for `hidTarget` so that you avoid goto in this function. I´ve just moved this wrapper to vfeature.h to ease usage.

ManOnTheMountainTech commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,268 @@
+
+    if (NT_SUCCESS(status)) {
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+        InterlockedIncrement((PLONG)(&pDeviceContext->fSetBlackSuccess));
```
Here's the rationale. Let's see if it's valid first. The IoRegisterPlugPlayNotification callback runs in one thread. CreateWorkItemForIoTargetOpenDevice runs in another. These two threads could be running on different processors. Microcode could be halfway through an operation on a processor core. In addition, the samples use Interlocked operations whenever two or more threads are doing a accessing a piece of memory.

forderud replied:
Please then add a comment to document that IoRegisterPlugPlayNotification & CreateWorkItemForIoTargetOpenDevice can be called simultaneously.

Also, I suggest replacing `InterlockedIncrement` with `InterlockedExchange` since it seems like you're really setting a flag here, and not incrementing a counter. Or am I misunderstanding your code?

forderud replied:
The `InterlockedIncrement` function expect a 32bit target pointer. The `fSetBlackSuccess` is of type `BOOLEAN` which appear to be a typedef for the 8bit  `UCHAR` type.

Won't the `InterlockedExchange` call above therefore lead to out-of-bounds write, which would corrupt the driver and possibly trigger a security vulnerability?

forderud commented on TailLight/device.cpp:
```diff
@@ -1,46 +1,83 @@
+            // NOTE: It is possible for us to get blocked waiting for a system
+            // thread. One solution would be to use a timer that spawns a
+            // system thread on timeout vs. a wait loop.
+            // Always must return success, so ignore result.
```
Please clarify what is meant by "Always must return success" here. Are you referring to `CreateWorkItemForIoTargetOpenDevice` or `PnpNotifyDeviceInterfaceChange`? If the latter, then where is the "must return success" requirement documented?

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,270 @@
+        }
+
+        // TODO: Wondering if we just cant cast pInBuffer as a TailLightReport
+        RtlCopyMemory(pInBuffer, &report, sizeof(TailLightReport));
```
You can avoid the RtlCopyMemory by transforming the code into:
```
TailLightReport* pInBuffer = nullptr;
// WdfMemoryCreate call...
{
    TailLightReport  report = {};
    report.Blue = 0x0;
    report.Green = 0x0;
    report.Red = 0x0;

    *pInBuffer = report;
}
```

forderud commented on TailLight/SetTaillightBlack.cpp:
```diff
@@ -0,0 +1,269 @@
+    WDFDEVICE device = static_cast<WDFDEVICE>(WdfWorkItemGetParentObject(workItem));
+    auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(workItem);
+
+    status = SetBlackAsync(device, workItemContext->symLink);
```
Please add logging if the `SetBlackAsync` call fails.

I'm observing >10 failing  `SetBlackAsync` calls when attaching a mouse when testing this PR. Still, the tail-light is cleared, so the function "works" from a functionality point-of-view. Still, these errors will need to be fixed.

forderud replied:
It seems to me like `SetBlackAsync` might still fail silently without any logging if the `WdfRequestSend` call within is failing. Please ensure that all failure cases are logged properly, so that they do not go unnoticed.

forderud replied:
Thanks for addressing this comment.

forderud commented on TailLight/vfeature.h:
```diff
@@ -26,13 +26,13 @@ class WDFIOTARGET_Wrap {
 
 
-NTSTATUS SetFeatureColor (
+NTSTATUS SetFeatureColor(
```
Why are you introducing whitespace changes to this file?

These changes increases the size of the PR in a unnecessary way. This both increases the review burden and adds risk of regressions.

forderud replied:
Please don't resolve this comment before it have been properly addressed.

forderud commented on TailLight/vfeature.cpp:
```diff
@@ -32,9 +32,21 @@ NTSTATUS SetFeatureColor (
+
+Returns:
+
+    Whether everything succeeded or a step failed.
```
The argument and return-value documentation above is obvious. There's little point in adding boilerplate documentation like this. You just increase verbosity.

forderud commented on TailLight/driver.cpp:
```diff
@@ -28,13 +43,3 @@ NTSTATUS DriverEntry(
-
-/** Driver unload callback.
-    Used to perform operations that must take place before the driver is unloaded.  */
-VOID EvtDriverUnload(
```
Why are you removing the `EvtDriverUnload` callback this way? Can't you just leave the function as-is?

forderud replied:
Please don't resolve this comment before it have been properly addressed.

forderud commented on TailLight/driver.cpp:
```diff
@@ -1,18 +1,33 @@
+
+    Returns:
+
+        A success code if all steps worked, otherwise a failure code.
```
Why are you bundling documentation changes for `DriverEntry` with this PR? This increases the size of the PR in an unnecessary way.

The documentation added for arguments and return-value is furthermore obvious, so just increases verbosity without adding any value.

forderud commented on TailLight/device.cpp:
```diff
@@ -1,52 +1,131 @@
+
+    Information about or describing the device bound
+    to the IO target.
+--*/
```
You're not changing the `GetTargetPropertyString` function in this PR. Therefore, please don't introduce doc updates to the function in this PR.

You're free to propose doc. updates to the function, but that should be in a _separate_ PR.

forderud commented on TailLight/TailLight.h:
```diff
@@ -23,7 +23,7 @@ struct TailLightReport {
 
         if ((Unknown1 != 0xB2) || (Unknown2 != 0x03)) {
-            //KdPrint(("TailLight: TailLightReport: Unknown control Code 0x%x 0x%x\n", Unknown1, Unknown2));
+            KdPrint(("TailLight: TailLightReport: Unknown control Code 0x%x 0x%x\n", Unknown1, Unknown2));
```
Why are you uncommenting this line?

forderud commented:
Kernel log observed when attaching an IntelliMouse:
```
TailLight: DriverEntry - WDF version built on Apr  1 2024 02:24:37
TailLight: PdoName: \Device\000000ba
TailLight: SetBlackAsync: Device \Device\00000069 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000005c not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000057 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000060 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000005d not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000005e not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000061 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000007d not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000063 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000062 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000054 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000007c not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\000000b5 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000007e not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000006b not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000005f not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000055 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000006a not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000006c not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000081 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000007f not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000080 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000065 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000066 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000056 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\000000b7 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000082 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000067 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\000000b8 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000068 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\000000b9 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\00000064 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\0000005b not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\000000b4 not known to control the taillight so failing
TailLight: SetBlackAsync: Device \Device\000000b6 not known to control the taillight so failing
DimpHidAddDevice: Failed to open device \\?\HID#VID_045E&PID_082A&MI_00&Col01#7&bb7cab7&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030} (err=5).
DimpHidAddDevice: Failed to open device \\?\HID#VID_045E&PID_082A&MI_01&Col01#7&181f730b&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\KBD (err=5).
TailLight: SetFeatureFilter
TailLight: SetBlackCompletionRoutine WdfRequestSend status: 0x0
Entered KbfEvtIoDeviceControl
Enter IOCTL_KBFILTR_SET_KEYBOARD_FILTERS 
State: NotBlocking
m_ModifierKeyState == 0x0, m_blockedModifierKeyStateWhenLayoutIsSet == 0x0
Entered KbfEvtIoDeviceControl
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
TailLight: SetFeatureFilter
```

The 22 `TailLight: SetFeatureFilter` entries appear to be a preexisting condition, and _not_ a regression in this PR. The 35 `not known to control the taillight so failing` entries are new. They're not really indicative of a failure or problem, but rather low specificity for the `IoRegisterPlugPlayNotification` that subscribes to all `GUID_DEVINTERFACE_HID` events. I wonder if this can be somehow narrowed down to reduce callback "noise"(?)
